exonum/src/blockchain/mod.rs

```diff
@@ -256,24 +257,22 @@ impl Blockchain {
         proposer_id: ValidatorId,
         height: Height,
         tx_hashes: &[Hash],
-        pool: &BTreeMap<Hash, Box<Transaction>>,
     ) -> (Hash, Patch) {
         // Create fork
         let mut fork = self.fork();
 
         let block_hash = {
-            // Get last hash
+            // Get last hash.
             let last_hash = self.last_hash();
-            // Save & execute transactions
+            // Save & execute transactions.
             for (index, hash) in tx_hashes.iter().enumerate() {
-                let tx = pool.get(hash).expect(
-                    "BUG: Cannot find transaction in pool.",
-                );
-
-                execute_transaction(tx.as_ref(), height, index, &mut fork);
+                self.execute_transaction(*hash, height, index, &mut fork)
+                    // execute_transaction could fail,with invalid
+                    // Transaction that cannot be deserialized, or not found in pool.
```

Comment:
Execution could fail if the transaction cannot be deserialized or it isn't found.